### DISCUSSION TITLE
Making sure ‘cbc_cmd’ result is deterministic

### DISCUSSION THREAD

**Discussion by s4ever** [coin-or/Cbc#379] [Q&A]
Hi,

I am trying to solve a fairly complex MIP modelled in PuLP using cbc_cmd. I am using default parameters. I find that at times the results are not deterministic. I did some search in internet and I know one parameter that affect will be timelimit.

I have few questions here

1.     What are the parameters I should set to make sure the solution is deterministic and repeatable?
2.     Any other steps to be done for deterministic solution? Such as deleting the Lp file, or restarting the environment etc for each run?
3.     Why timelimit parameter should affect determinism for a smaller problem? For example, say if my problem is simple enough to solve well within the timelimit, why could each run have different results?

Thank you in advance for all your help!

**Comment by guifcoelho** [comment]
Does it make sense for you to run with a fixed seed? For instance, you can add the cmd line option "RandomC 1"

**Reply by s4ever** [reply]
I can try that. Do you know what’s being set by this random number? Branching strategy?

also, is there any documentation which lists the parameters and gives some details?

**Reply by guifcoelho** [reply]
The "RandomC" option does this:
```
randomC(bcSeed) : Random seed for Cbc
Allows initialization of the random seed for pseudo-random numbers
used in heuristics such as the Feasibility Pump to decide whether
to round up or down. The special value of 0 lets Cbc use the time
of the day for the initial seed.
<Range of values is -1 to 2147483647;
```

The most complete cheatsheet I've found so far was [this](http://www.decom.ufop.br/haroldo/files/cbcCommandLine.pdf).

Also, you can call the solver directly and enter "?". It will return all available options.

```
C:\Users\user>"C:\Users\user\cbc.exe"
Welcome to the CBC MILP Solver
Version: 2.10.3
Build Date: Dec 15 2019

CoinSolver takes input from arguments ( - switches to stdin)
Enter ? for list of commands or help
Coin:?
In argument list keywords have leading - , -stdin or just - switches to stdin
One command per line (and no -)
abcd? gives list of possibilities, if only one + explanation
abcd?? adds explanation, if only one fuller help
abcd without value (where expected) gives current value
abcd value sets value
Commands are:
Double parameters:
 dualB(ound) dualT(olerance) primalT(olerance) primalW(eight) psi
 zeroT(olerance)
...
```

If you want more details on a given option, you can enter the option name followed by "?" or "??":
```
Coin:dualB??
dualB(ound) : Initially algorithm acts as if no gap between bounds exceeds this value
The dual algorithm in Clp is a single phase algorithm as opposed to
a two phase algorithm where you first get feasible then optimal.
If a problem has both upper and lower bounds then it is trivial to
get dual feasible by setting non basic variables to correct bound.
If the gap between the upper and lower bounds of a variable is more
than the value of dualBound Clp introduces fake bounds so that it
can make the problem dual feasible.  This has the same effect as a
composite objective function in the primal algorithm.  Too high a
value may mean more iterations, while too low a bound means the code
may go all the way and then have to increase the bounds.  OSL had
a heuristic to adjust bounds, maybe we need that here.
<Range of values is 1e-20 to 1e+12;
        current 1e+10>
```

**Reply by s4ever** [reply]
Thank you so much for this! You're a star! 
Let me try fixing the RandomC and look at the other parameters!

**Reply by guifcoelho** [reply]
Glad I could help. Have fun.

**Reply by bsaunders23** [reply]
As a quick note, you should set the `RandomS(eed)` argument instead, `RandomC(bcSeed)` actually defaults to a static seed value of -1. See example in https://github.com/coin-or/pulp/issues/545#issuecomment-1355737609

**Reply by behrisch** [reply]
And another note: setting the PYTHONHASHSEED as recommended in the [same discussion above](https://github.com/coin-or/pulp/issues/545#issuecomment-1355737609) is a good idea as well
